src/abstract-interpretation/processor.ts

```diff
@@ -1,7 +1,157 @@
+	return new Set([{min: n, minInclusive: true, max: n, maxInclusive: true}])
+}
+
+function getIntervalsOfDfgChild(node: NodeId, dfg: DataflowInformation): Intervals {
+	const dfgNode: [DataflowGraphVertexInfo, OutgoingEdges] | undefined = dfg.graph.get(node)
+	guard(dfgNode !== undefined, `No DFG-Node found with ID ${node}`)
+	const [_, children] = dfgNode
+	const ids = Array.from(children.entries())
+		.filter(([_, edge]) => edge.types.has(EdgeType.Reads))
+		.map(([id, _]) => id)
+	if(ids.length === 0) {
+		return new Set()
+	}
+	guard(ids.length === 1, `Multiple reads-edges found for node ${node}`)
+	const id = ids[0]
+	const constraint = constraintMap.get(id)
+	guard(constraint !== undefined, `No constraint found for ID ${id}`)
+	return constraint.intervals
+}
+
+class Assignment implements IHandler<Constraints> {
+	private lhs:           NodeId | undefined
+	private rhs:           NodeId | undefined
+	private readonly node: RBinaryOp<ParentInformation>
+	name = 'Assignment'
+
+	constructor(node: RBinaryOp<ParentInformation>) {
+		this.node = node
+	}
+
+	enter(): void {
+		console.log(`Entered ${this.name} ${this.node.info.id}`)
+	}
+
+	exit(): Constraints {
+		console.log(`Exited ${this.name} ${this.node.info.id}`)
+		return {
+			node:      this.lhs as NodeId,
+			intervals: new Set(), // TODO: check interval of the rhs
+			debugMsg:  this.name
+		}
+	}
+
+	next(constraint: Constraints): void {
+		if(this.lhs === undefined) {
+			this.lhs = constraint.node
+		} else if(this.rhs === undefined){
+			this.rhs = constraint.node
+		}
+		console.log(`${this.name} received ${constraint.debugMsg}`)
+	}
+}
+
+class BinOp implements IHandler<Constraints> {
+	private readonly node: RBinaryOp<ParentInformation>
+
+	constructor(node: RBinaryOp<ParentInformation>) {
+		this.node = node
+	}
+
+	name = 'Bin Op'
+
+	enter(): void {
+		console.log(`Entered ${this.name}`)
+	}
+
+	exit(): Constraints {
+		console.log(`Exited ${this.name}`)
+		return {
+			node:      this.node.info.id,
+			intervals: new Set(), // TODO: Check the operands constraints and see how the operation affects those
+			debugMsg:  this.name
+		}
+	}
+
+	next(value: Constraints): void {
+		console.log(`${this.name} received ${value.debugMsg}`)
+	}
+}
 
 export function runAbstractInterpretation(ast: NormalizedAst, dfg: DataflowInformation): DataflowInformation {
-	console.log('hi from abstract interpretation')
+	const cfg = extractCFG(ast)
+	const operationStack = new Stack<IHandler<Constraints>>()
+	visitCfg(cfg, (node, _) => {
+		const astNode = ast.idMap.get(node.id)
+		// TODO: avoid if-else
+		if(astNode?.type === RType.BinaryOp) {
+			switch(astNode.flavor) {
+				case 'assignment': operationStack.push(new Assignment(astNode)).enter(); break
+				case 'arithmetic': operationStack.push(new BinOp(astNode)).enter; break
+				default: guard(false, `Unknown binary op ${astNode.flavor}`)
+			}
+		} else if(astNode?.type === RType.Symbol) {
+			operationStack.peek()?.next({
```

Comment:
we should robustify this. What if the program does have no assignment, binop, ...
For example, if its just `x`. While simply ignoring it is correct in this scenario, it is a potential problem with implicit returns and function definitions (again, dataflow v2 should replace those with explicit returns).